[CASSANDRA-16325] Update streaming metrics incrementally

isaacreath opened the pull request:
[CASSANDRA-16325] Update streaming metrics incrementally

Currently the inbound and outbound streamed bytes metrics are incremented after each file is streamed, what doesn't represent the current number of bytes streamed since it can take a long time for a large file to be streamed. This patch changes the `StreamSession` class to increment the total bytes streamed every time the `StreamSesson` makes progress instead of when a file finishes streaming. 

patch by Isaac Reath and Dejan Gvozdenac ; reviewed by Paulo Motta for CASSANDRA-16325

Co-authored-by: Isaac Reath
Co-authored-by: Dejan Gvozdenac


The [Cassandra Jira](https://issues.apache.org/jira/projects/CASSANDRA/issues/)

isaacreath committed:
[CASSANDRA-16325] Update streaming metrics incrementally
- Add StreamMetricsHandler class which is a StreamEventHandler that processes ProgressEvent objects.
- Fix bug in CassandraCompressedStreamReader where we were over-reporting deltas between calls to StreamSession::progress.
- Add StreamMetricsHandler on StreamManager initialization.
- Add testing for incremental metrics:
	- Update cluster size to 2 and RF=2 to make it easier to reason about state
	- Create single 10MB sstable on node1 with compression disabled
	- Check that more than 3 incoming/outgoing bytes
          are reported during streaming from node1 to node2
	- Check that final transferred bytes matches expected sstable size (10MB)

isaacreath commented on test/distributed/org/apache/cassandra/distributed/test/metrics/StreamingMetricsTest.java:
```diff
@@ -145,6 +151,92 @@ public void testMetricsWithRepairAndStreamingToTwoNodes() throws Exception
+        ExecutorService nodetoolExecutor = new ThreadPoolExecutor(1, 1, 0L, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<>());
+        checkThatNoStreamingOccuredBetweenTheThreeNodes(cluster);
+        Future<Integer> streamingOperationExecution = nodetoolExecutor.submit(streamingOperation);
+        Thread.sleep(500); // Wait for some streaming to happen.
```
If we could find a way to get a notification instead of sleep that would be nice.

smiklosovic commented on src/java/org/apache/cassandra/streaming/StreamSession.java:
```diff
@@ -1027,9 +1022,30 @@ public void receive(IncomingStreamMessage message)
+    private void updateMetricsOnProgress(ProgressInfo progress)
+    {
+        ProgressInfo.Direction direction = progress.direction;
+        long lastSeenBytesStreamedForProgress = lastSeenBytesStreamed.getOrDefault(progress, 0L);
```
does this mean that `lastSeenBytesStreamed` map will be ever growing only as I dont see any place where we are removing it from that map?

smiklosovic commented on src/java/org/apache/cassandra/streaming/StreamSession.java:
```diff
@@ -1027,9 +1022,30 @@ public void receive(IncomingStreamMessage message)
+            StreamingMetrics.totalOutgoingBytes.inc(newBytesStreamed);
+            metrics.outgoingBytes.inc(newBytesStreamed);
+        }
+
```
nit: empty line redundant

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/metrics/StreamingMetricsTest.java:
```diff
@@ -145,10 +155,159 @@ public void testMetricsWithRepairAndStreamingToTwoNodes() throws Exception
+    @Test
+    public void testMetricsUpdateIncrementallyWithRepairAndStreamingBetweenNodes() throws Exception
+    {
+        try(Cluster cluster = init(Cluster.build(2)
```
nit: `try (Cluster` - there should be space.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/metrics/StreamingMetricsTest.java:
```diff
@@ -145,10 +155,159 @@ public void testMetricsWithRepairAndStreamingToTwoNodes() throws Exception
+    @Test
+    public void testMetricsUpdateIncrementallyWithRebuildAndStreamingBetweenNodes() throws Exception
+    {
+        try(Cluster cluster = init(Cluster.build(2)
```
nit: `try (Cluster` - there should be space.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/metrics/StreamingMetricsTest.java:
```diff
@@ -145,10 +155,159 @@ public void testMetricsWithRepairAndStreamingToTwoNodes() throws Exception
+        cluster.schemaChange(String.format("CREATE TABLE %s.cf (k text PRIMARY KEY, c1 text) " +
+                                           "WITH compaction = {'class': '%s', 'enabled': 'false'} " +
+                                           "AND compression = {'enabled':'false'};",
+                                           KEYSPACE, "SizeTieredCompactionStrategy"));
```
I fail to see why there is a placeholder for `class` when it is a constant. Why is it extracted?

isaacreath replied:
This just keeps consistent with the way we create a schema in the existing tests (see: https://github.com/apache/cassandra/blob/trunk/test/distributed/org/apache/cassandra/distributed/test/metrics/StreamingMetricsTest.java#L71). If it's better to be simple, we change it in both places.

smiklosovic commented on test/distributed/org/apache/cassandra/distributed/test/metrics/StreamingMetricsTest.java:
```diff
@@ -145,10 +155,159 @@ public void testMetricsWithRepairAndStreamingToTwoNodes() throws Exception
+        // Disable dropping of messages from node1 to node2
+        drop1to2.off();
+
+        ExecutorService nodetoolExecutor = new ThreadPoolExecutor(1, 1, 0L, TimeUnit.MILLISECONDS, new LinkedBlockingQueue<>());
```
should not we shutdown this executor after we are done?

smiklosovic commented on src/java/org/apache/cassandra/streaming/StreamSession.java:
```diff
@@ -1027,9 +1022,30 @@ public void receive(IncomingStreamMessage message)
+            metrics.incomingBytes.inc(newBytesStreamed);
+        }
+
+        lastSeenBytesStreamed.put(progress, lastSeenBytesStreamedForProgress + newBytesStreamed);
```
here we are just adding (presumably new) `progress`-es again and again. What does this mean in practice? If we are still adding the very same `progress` object, why does it have to be a map? Hence, I guess `progress` _is_ new every now and then. So the underlying map grows beyond any limit until streaming session is finished? What if we have thousands streaming sessions at once?

isaacreath replied:
In practice we will be adding a new progress object for every file streamed in each direction by this `StreamSession` (see: [ProgressInfo::hashCode](https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/streaming/ProgressInfo.java#L98)). The worst case is the number of total entries is equal to number of files on the local node * number of files on the remote node for each `StreamSession`. 

A simple optimization I can add here would be to remove the progress object from the map once we've completed streaming. In this case when `progress.currentBytes == progress.totalBytes`.  This would clean things up as each file completes and probably improve memory utilization in the average case. This wouldn't handle the worst case where all files complete streaming at the same time.

smiklosovic replied:
_ The worst case is the number of total entries is equal to number of files on the local node * number of files on the remote node for each StreamSession._

Is it really _times_ (multiplication)? So thousand files on each side is 1 million progress objects? Or you meant _plus_ ?

Anyway I guess removing it would help indeed, we probably need to be cautions what to do when a streaming session fails in the middle and it is just garbage collected or something similar ... 


btw why do we need to hold whole progress objects? the only thing needed is `currentBytes` from it. If there is something unique in `ProgressInfo`, why not to use it as a key in that map?

isaacreath replied:
Good catch, yeah `+`.

smiklosovic replied:
also, would it be enough to pass just `bytes` used upon ProgressInfo instantiation into `updateMetricsOnProgress` method as another argument? So you would not need to bypass that object at all, probably. I am just trying to find some workaround.

smiklosovic replied:
I think you ultimately need just file name, direction and bytes and you might probably just hash file name and direction to have it unique but I am just figuring this out in my mind, not tested nor anything ... 

Hashing just these two things is probably not necessary either if you dedicate two separate maps for each direction so key in each map would be just a file name.

smiklosovic replied:
````
        
        Map<String, long[]> ins = new HashMap<>();
        Map<String, long[]> outs = new HashMap<>();

        // at the beginning we would put it in there based on direction        
        ins.put(progress.fileName, new long[] {0, progress.totalBytes});
````

So later we could check when to remove it from that map as we have access to `totalBytes` we previously put there.

isaacreath replied:
Yes, what we really care about to identify the uniqueness is the `(fileName, direction)` tuple, so two maps could work.

I'm not really sure about why you're looking to store the `progress.totalBytes` value in the `ins` and `outs` maps.  The call to `StreamSession::progress` _should_ always set the same value for a given `(fileName, direction)` pair.

smiklosovic replied:
too hard to elaborate on that here and now, lets see what you come up with and then we will see

smiklosovic replied:
I think that main reason for putting there totalBytes was to know when you are allowed to remove that entry from the map based on your `progress.currentBytes == progress.totalBytes` suggestion.

we should really double check what happens when a streaming session fails - what happens with this whole class and how it behaves under such circumstances.

isaacreath replied:
I tried out removing from the map when `progress.currentBytes == progress.totalBytes` and found that this function is called in [CassandraCompressedStreamReader::read](https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/db/streaming/CassandraCompressedStreamReader.java#L96) multiple times (seemingly once per partition written as a file). What this results in is `progress` being called multiple times where `currentBytes == totalBytes`.  I'm guessing this is a behavior we'd like to keep instead of moving the call to progress outside of the `while` loop in `CassandraCompressedStreamReader::read`.

Given that, I'm going to spend some time thinking of another way to clean up this map when file streaming is done.

smiklosovic commented on src/java/org/apache/cassandra/streaming/StreamSession.java:
```diff
@@ -1027,9 +1022,30 @@ public void receive(IncomingStreamMessage message)
+            metrics.outgoingBytes.inc(newBytesStreamed);
+        }
+
+        else if (direction == ProgressInfo.Direction.IN)
```
ProgressInfo.Direction can be only IN or OUT. If it is not OUT, it is IN, so we do not need `else if`, is not `else` good enough? Or direction can be null?  I would `assert` that direction is not null and then the `else` branch does not need `else if`.

isaacreath replied:
In theory `direction` can be null, but in practice, it doesn't appear to be (all calls to `StreamSession::progress` seem to set a value).

Instead of adding the assert, what about using the `@Nonnull` annotation to prevent anyone from passing null in the future. Then we can just use `else`.

isaacreath commented:
@smiklosovic After thinking about this issue for a little bit, @pauloricardomg  and I came up with a different approach. 

We moved the logic which increments the incoming / outgoing bytes metrics from the `StreamSession` to a dedicated class called the `FileStreamMetricsListener`. `FileStreamMetricsListener` objects will then be created by each `CassandraIncomingFile` and `CassandraOutgoingFile`. These then pass the `FileStreamMetricsListener` to the appropriate stream reader / writer. As those reader / writers transfer bytes, `FileStreamMetricsListener::onBytesTransferred` will be called to update the metrics. When file transfer is complete, we call `FileStreamMetricsListener::onStreamSuccessful` to verify that we've tracked the size appropriately (thus ensuring that any new stream reader / writers are calling `onBytesTransferred`). 

This approach addresses the concerns of having an ever-growing map of `ProgressInfo` objects and it neatly encapsulates the logic for updating these counter objects. There are some aspects of this approach of which I am not a fan, namely the fact that there is little ensuring that new Stream reader / writer objects actually create a `FileStreamMetricsListener` and update it properly; but, from looking at the code, there is some refactoring which needs to be done to unify all of the stream reader / writer classes in a single hierarchy that should probably be handled outside the scope of this ticket before we could consider cleaning up this approach.

smiklosovic commented on src/java/org/apache/cassandra/db/streaming/FileStreamMetricsListener.java:
```diff
@@ -0,0 +1,96 @@
+ */
+public class FileStreamMetricsListener
+{
+    public static final FileStreamMetricsListener NO_OP = new FileStreamMetricsListener(0L, FBUtilities.getLocalAddressAndPort(), ProgressInfo.Direction.IN){
```
Should not be the curly bracket at the new line too? There is also a space missing `){` vs `) {`. 

It is also a little bit counter-intuitive to have a direction `IN` used for NO_OP listener. Why not `OUT`? I think it would be preferrable to define a `Direction` called `NO_OP` or `NA` (Not Available) for this purpose.

However, by calling this constructor, we are running through the logic which interacts with `StreamingMetrics` etc. I do not think that is necessary at all. We could create an empty constructor for `FileStreamMetricsListener` and initialise class properties with nulls for counters and 0 for totalSize in it. That way we avoid messing with `Direction` too.

That no-arg constructor should be private.

isaacreath replied:
Agree that is much cleaner.

pauloricardomg replied:
I was wondering if we should make `FileStreamMetricsListener NO_OP` an explicit testing method instead of a static field (ie. `static FileStreamMetricsListener forTesting()` ) to indicate this is only valid for testing and should not be used by db code - wdyt?

smiklosovic replied:
It should be package protected and annotation `@VisibleForTesting` should be added on that method.

However, creating a method for that seems to be like unnecessary exercise. Nothing wrong with having it like it is, package protected and annotated as we usually do. That serves the same purpose as having "forTesting()" method.

pauloricardomg replied:
SGTM 👍

pauloricardomg replied:
hmm actually some tests outside the package are using `NO_OP`, so it will probably need to remain public.

smiklosovic replied:
ah right, that is fine if we remain to have that annotation on it I guess?

smiklosovic commented on src/java/org/apache/cassandra/db/streaming/FileStreamMetricsListener.java:
```diff
@@ -0,0 +1,96 @@
+    private final Counter peerOutgoingBytesStreamed;
+    private long lastSeenBytes = 0L;
+
+    public FileStreamMetricsListener(long totalSize, InetAddressAndPort peer, ProgressInfo.Direction direction)
```
non null annotation for `direction` maybe as it was done previously? If we introduce `Direction.NA` then `else` would need to be changed to `else if` for `IN`.

smiklosovic commented on src/java/org/apache/cassandra/db/streaming/FileStreamMetricsListener.java:
```diff
@@ -0,0 +1,96 @@
+            this.totalOutgoingBytesStreamed = StreamingMetrics.totalIncomingBytes;
+            this.peerOutgoingBytesStreamed = StreamingMetrics.get(peer).incomingBytes;
+        }
+
```
nit: redundant empty line

smiklosovic commented on src/java/org/apache/cassandra/db/streaming/CassandraStreamReader.java:
```diff
@@ -109,7 +111,6 @@ public CassandraStreamReader(StreamMessageHeader header, CassandraStreamHeader s
     public SSTableMultiWriter read(DataInputPlus inputPlus) throws Throwable
     {
         long totalSize = totalSize();
-
```
just leave it as it was?

smiklosovic replied:
@isaacreath maybe you forgot this?

smiklosovic commented on src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java:
```diff
@@ -172,10 +174,11 @@ public void write(StreamSession session, StreamingDataOutputPlus out, int versio
+                                           new CassandraStreamWriter(sstable, header, session, fileStreamMetricsListener);
             writer.write(out);
         }
+        fileStreamMetricsListener.onStreamSuccessful();
```
what if `onStreamSuccessful` returned boolean so it would be called `isStreamSuccessful` and if it is not it would throw an exception? This method already throws `IOException` so it seems to be quite logical to me that if streaming was not successful we would throw as well. Right now we are doing `assert` in it which can throw `AssertionError` but that is not subclass of `IOException`.

Another option is to throw IOException directly from `onStreamSuccessful` with accompanying error message but that feels strange.

pauloricardomg replied:
I don't think `FileStreamMetricsListener` should be responsible for determining whether a stream is successful, since this class is meant to be a dumb consumer. For instance, I think we should update the number of incoming files streamed on this method `(metrics.countStreamedIn(isEntireSSTable))`, since we can assume the stream is completed successfully when `onStreamSuccessful()` is called. The assertion `assert lastSeenBytes == totalSize;` on `onStreamSuccessful()` is just to verify that the behavior of the class `FileStreamMetricsListener` is correct, not to check whether a stream is successful or not.

isaacreath replied:
I tried making the change to `isStreamSuccessful` from `onStreamSuccessful` and it seems to make the `FileStreamMetricsListener` a very leaky abstraction to provide a proper error message and we end up duplicating the checking logic between the `CassandraIncomingFile` and `CassandraOutgoingFile` which I'm not a huge fan of (and again, seems like a leaky abstraction).

I could instead throw an `IOException` instead of running the assert; but, like you say, that feels strange as no IO is actually happening in `onStreamSuccessful`.

smiklosovic replied:
Yes I get that but code-wise that method might throw AssertionError but the method that method is called in throws IOException and that is (technically) just wrong.

pauloricardomg replied:
> that method might throw AssertionError but the method that method is called in throws IOException and that is (technically) just wrong.

I don't see what is wrong about this, as far as I understand assertions can be added anywhere in the code to validate assumptions, without requiring methods to throw `AssertionError` explicitly. Am I missing something?

smiklosovic replied:
OK Paulo.

May I ask, what happens when it actually throws that exception (AssertionError). What happens in runtime? Where it propagates and what eventually fails?

Just honestly asking as I have never checked out this code in IDE to dig deeper, sorry ...

pauloricardomg replied:
> May I ask, what happens when it actually throws that exception (AssertionError). What happens in runtime? Where it propagates and what eventually fails?

The way I see it is that assertions shouldn't happen in runtime unless the code is broken, as they're errors targeted for developers to detect broken code assumptions. Unfortunately in Cassandra we overload assertions a bit and use it for error handling (which IMO is a bad smell), but this is not what is being done here.

The assertion `assert lastSeenBytes == totalSize` on `FileStreamMetricsListener.onStreamSuccessful` is a runtime safety check to ensure that the behavior of the class `FileStreamMetricsListener` is working as expected and we don't report more streamed bytes than the expected amount. It's a future-proof safety check for programming errors, but should not be expected to fail during correct operation.  If this assertion is ever thrown, it means that there is a bug in the code that must be fixed.

isaacreath replied:
For a sanity check, we can look into whether or not this can happen during valid operation before we commit this assertion. As far as I can tell it can't, but it's worth digging into further.

smiklosovic commented on src/java/org/apache/cassandra/db/streaming/CassandraEntireSSTableStreamReader.java:
```diff
@@ -95,7 +97,6 @@ public SSTableMultiWriter read(DataInputPlus in) throws Throwable
         ComponentManifest manifest = header.componentManifest;
         long totalSize = manifest.totalSize();
-
         logger.debug("[Stream #{}] Started receiving sstable #{} from {}, size = {}, table = {}",
```
probably leave it as it was?

smiklosovic commented on src/java/org/apache/cassandra/db/streaming/CassandraCompressedStreamReader.java:
```diff
@@ -57,7 +57,6 @@ public CassandraCompressedStreamReader(StreamMessageHeader header, CassandraStre
     public SSTableMultiWriter read(DataInputPlus inputPlus) throws Throwable
     {
         long totalSize = totalSize();
-
```
probably leave it as it was?

smiklosovic commented on src/java/org/apache/cassandra/streaming/StreamSession.java:
```diff
@@ -200,6 +203,8 @@ public class StreamSession implements IEndpointStateChangeSubscriber
 
     private final TimeUUID pendingRepair;
     private final PreviewKind previewKind;
+    private final Map<String, Long> lastSeenIncomingBytesStreamed;
```
this does not seem to be necessary anymore with all other changes in this class?

smiklosovic commented on src/java/org/apache/cassandra/streaming/StreamSession.java:
```diff
@@ -36,13 +36,16 @@
 import java.util.function.Function;
 import java.util.stream.Collectors;
 
+import javax.annotation.Nonnull;
```
check unused imports maybe?

smiklosovic reviewed (commented):
Thanks @isaacreath . I left couple comments again.

smiklosovic commented on src/java/org/apache/cassandra/db/streaming/FileStreamMetricsListener.java:
```diff
@@ -0,0 +1,113 @@
+    /**
+     * Used for NO-OP listener.
+     */
+    private FileStreamMetricsListener(){
```
@isaacreath put `{` on new line please.

smiklosovic commented on src/java/org/apache/cassandra/db/streaming/FileStreamMetricsListener.java:
```diff
@@ -0,0 +1,113 @@
+    {
+        assert lastSeenBytes == totalSize;
+    }
+
```
@isaacreath empty line not necessary.

smiklosovic commented on src/java/org/apache/cassandra/db/streaming/FileStreamMetricsListener.java:
```diff
@@ -0,0 +1,113 @@
+    }
+
+   @VisibleForTesting
+   public FileStreamMetricsListener(long totalSize, Counter totalOutgoingBytesStreamed, Counter peerOutgoingBytesStreamed)
```
@isaacreath just checking if this needs to be explicitly `public` or it may be package protected.

isaacreath replied:
I can't imagine any reason it needs to be used outside of the specific unit tests for this class. Removed it.

smiklosovic commented on src/java/org/apache/cassandra/db/streaming/FileStreamMetricsListener.java:
```diff
@@ -0,0 +1,113 @@
+        this.totalSize = totalSize;
+        if (direction == ProgressInfo.Direction.OUT)
+        {
+            this.totalOutgoingBytesStreamed = StreamingMetrics.totalOutgoingBytes;
```
@isaacreath  I am not big fan of `this`. There is nothing but _this_ `totalOutgoingBytesStreamed`, it can not happen we need `super` one etc ... I think `this` is just redundant here, same for all other cases in constructors.

This class does not extend anything either so we would need to specifically refer to `this`.

isaacreath replied:
Removed extraneous `this` usage.

smiklosovic commented on src/java/org/apache/cassandra/db/streaming/FileStreamMetricsListener.java:
```diff
@@ -0,0 +1,113 @@
+    {
+        long delta = cumulativeBytesTransferred - lastSeenBytes;
+        assert delta >= 0;
+        this.totalOutgoingBytesStreamed.inc(delta);
```
@isaacreath why `this` ?

isaacreath replied:
Removed extraneous `this` usage.

smiklosovic commented on src/java/org/apache/cassandra/db/streaming/FileStreamMetricsListener.java:
```diff
@@ -0,0 +1,113 @@
+
+    public FileStreamMetricsListener(long totalSize, InetAddressAndPort peer, @Nonnull ProgressInfo.Direction direction)
+    {
+        this.totalSize = totalSize;
```
in this case, `this` is indeed justified.

smiklosovic commented on src/java/org/apache/cassandra/db/streaming/FileStreamMetricsListener.java:
```diff
@@ -0,0 +1,113 @@
+     */
+    private FileStreamMetricsListener()
+    {
+        this.totalSize = 0L;
```
@isaacreath `this` in this constructor is not necessary either.

smiklosovic commented on test/microbench/org/apache/cassandra/test/microbench/ZeroCopyStreamingBenchmark.java:
```diff
@@ -145,7 +145,6 @@ public void setupBenchmark() throws IOException
                                      .withFirstKey(sstable.first)
                                      .withTableId(sstable.metadata().id)
                                      .build();
-
```
@isaacreath I know it is super annoying but if you manage to put these lines back (on line 171 too), this file would stayed unchanged and your patch would be smaller (in total 3 files less when all my comments are reflected!) I totally get that this seems like a lot of red tape but we need to contain the dragon / lower the entropy as much as we can.

smiklosovic replied:
cc @pauloricardomg

smiklosovic commented on src/java/org/apache/cassandra/db/streaming/CassandraOutgoingFile.java:
```diff
@@ -170,7 +170,6 @@ public void write(StreamSession session, StreamingDataOutputPlus out, int versio
             // legacy streaming is not affected by stats metadata mutation and index sumary redistribution
             CassandraStreamHeader.serializer.serialize(header, out, version);
             out.flush();
-
```
@isaacreath this change is not necessary, please move it back so this file is not marked as changed in git.

smiklosovic commented on src/java/org/apache/cassandra/db/streaming/CassandraIncomingFile.java:
```diff
@@ -68,7 +68,6 @@ public synchronized void read(DataInputPlus in, int version) throws Throwable
         CassandraStreamHeader streamHeader = CassandraStreamHeader.serializer.deserialize(in, version);
         logger.debug("Incoming stream entireSSTable={} components={}", streamHeader.isEntireSSTable, streamHeader.componentManifest);
         session.countStreamedIn(streamHeader.isEntireSSTable);
-
```
@isaacreath same

isaacreath committed:
make sure to make CassandraIncomingFile, CassandraOutgoingFile and ZeroCopyStreamingBenchmark untouched

isaacreath committed:
clean up whitespace in CassandraOutgoingFile

isaacreath committed:
separate streaming tests into their own classes, create utility class for running test logic.

isaacreath committed:
create cluster object in all streaming metrics test cases

isaacreath committed:
Move streaming metrics to two nodes tests into their own classes. Make the StreamingMetricsTestUtils not inherit from TestBaseImpl

isaacreath committed:
log bytes we are adding to counter to verify tests

isaacreath committed:
allow us to specify the number of SSTables to stream, reduce streaming throughput

isaacreath committed:
remove unused import in StreamingMetricsTestUtils
